### ISSUE TITLE
Option for builder to initialise collection fields to null unless a builder method is explicitly called

### ISSUE THREAD

**Issue by glyn** [immutables/immutables#299]
Our use case is that we are providing a rich Java API ([1]) to drive a correspondingly rich REST API. In the REST API, there are certain "update" operations where an empty collection means "update the collection to empty" and a null collection means "do not update the collection". Essentially, parameters which are not set via the builder need to be left "unset" so the update can be applied only to the parameters which were explicitly set. So there is a semantic distinction between null and an empty collection.

The solution we are requesting ([2]) for Lombok is to allow a null initial collection to be specified, but keep an empty initial collection as the default.

This issue is for Immutables to provide a similar solution. See [3] for the full thread which led to this issue.

Towards the end of the thread, @elucash wrote:

"I think it might be possible to make @Nullable collections to have "add" and "addAll" sort of methods generated on a builder. In this case, the default, unset value would be null (to be consistent with other nullable values), and you can then set it to empty collection using explicit initializer `.setList(Collections.emptyList())`.".

I suspect the default behaviour should be for collection fields to be initialised with empty collections and for the new behaviour requested by this issue to be made optional in some way.

[1] https://github.com/cloudfoundry/cf-java-client
[2] https://github.com/rzwitserloot/lombok/issues/1071
[3] https://groups.google.com/d/msg/immutables/nv0VKt95iUE/xYxvBfM9IwAJ

**Comment by elucash** [comment]
Thank you for the feature request!

Regarding 
> I suspect the default behaviour should be for collection fields to be initialised with empty collections and for the new behaviour requested by this issue to be made optional in some way.

Here might be two interconnected features that might worth mentioning.

1. Support for convenient initialization (`add/addAll`) of `@Nullable` collections. The default value is clearly should be `null` as for any nullable attribute.

```java
@Nullable List<String> getNames();
...builder()
  .addNames("Nameless")  // will initialize collection with values
  .addAllNames(Arrays.asList("Nameful", "Nameless"))
  ...
  .setNames(Collection.emptyList()) initialize to empty collection
  .setNames(null) // initialize to null explicitly (null is default if initializer omitted)
  ...
  .build();
```

2. Support for convenient initialization  (`add/addAll`) of `@Value.Default` collections. But given the fact that default attributes might also be `@Nullable` it can be both null or empty (or whatever provided by initializer method) collection by default. It does provide syntactic overhead

```java
// same as the example above, but provides default value explicitly
@Value.Default @Nullable List<String> getNames() {
   return ImmutableList.of();
}
```

That was my attempt to shoehorn the expected result into what I think can work and make sense for Immutables. Little do we desire to introduce an ad-hoc knob for "empty by default, but can be set null" kind of collection attribute.

**Comment by elucash** [comment]
2.2-beta released with support for nullable/default collections as described above, inviting to evaluate it

**Comment by glyn** [comment]
Thanks @elucash! I'm no longer working on the cf-java-client, but I've notified the team about this feature in case they have comments.

**Comment by nebhale** [comment]
@elucash I'm taking over where @glyn and will take a look this morning.  Thanks for the quick turn around.

**Comment by elucash** [comment]
released in 2.2 Thank you! Feel free to open additional issues or follow up here
